packages/compiler-core/src/transforms/vOn.ts

```diff
@@ -120,7 +120,11 @@ export const transformOn: DirectiveTransform = (
       // avoiding the need to be patched.
       if (shouldCache && isMemberExp) {
         if (exp.type === NodeTypes.SIMPLE_EXPRESSION) {
-          exp.content = `${exp.content} && ${exp.content}(...args)`
+          if (exp.content.startsWith('$setup')) {
+            exp.content = `${exp.content} && ${exp.content}.call(undefined, ...args)`
```

Comment:
Is this change necessary? Can you add a test for it?

Reply:
I see. I still wonder if these changes shouldn't be only applied in dev as I imagine that we want both the behavior to be the same in prod/dev but also we don't want to unnecessarily make the generated code heavier in production

Reply:
> I see. I still wonder if these changes shouldn't be only applied in dev as I imagine that we want both the behavior to be the same in prod/dev but also we don't want to unnecessarily make the generated code heavier in production

Yeah. $setup only exists in the development environment. This change will not affect compilation in the production environment.